### PR TITLE
Support for passing multiple separated values with one param

### PR DESCRIPTION
**Pull request by jradlica**
Related to issues: [#32 ](https://github.com/tkaczmarzyk/specification-arg-resolver/issues/32), [#63 ](https://github.com/tkaczmarzyk/specification-arg-resolver/issues/63)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,17 @@
+UNRELEASED
+======
+
+* Added support for passing multiple separated values with one param. Added `paramSeparator` to `@Spec` interface, which allows for specifying regex which is used to the delimitation of param value. For example controler method:
```
Comment (tkaczmarzyk) [inline CHANGELOG.md:4]: Added support for passing multiple values as a single HTTP parameter. Then new `paramSeparator` attribute of `@Spec` can be used to define the separator (e.g. comma). For example the following controller method:

Comment (tkaczmarzyk) [reply CHANGELOG.md:4]: Additionally:
- let's start with simple char/string separators, without regex support -- regexp might become dangerous in some cases, and I think that comma will suffice for most use-cases

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,17 @@
...
+```java
+@RequestMapping(value = "/customers", params = "genderIn")
+@ResponseBody
+public Object findCustomersByGender(
+	@Spec(path = "gender", params = "genderIn", paramSeparator = ",", spec = In.class) Specification<Customer> spec) {
+	return customerRepo.findAll(spec);
+}
+```
+handle request `GET http://myhost/customers?gender=MALE,FEMALE` in the same way what request `GET http://myhost/customers?gender=MALE?gender=FEMALE`
```
Comment (tkaczmarzyk) [inline CHANGELOG.md:13]: will handle `GET ...` in exactly the same way as `GET ....`

File: README.md [hunk 3]
```diff
@@ -91,6 +91,8 @@ Simple specifications
 
 Use `@Spec` annotation to automatically resolve a `Specification` argument of your controller method. `@Spec` has `path` property that should be used to specify property path of the attribute of an entity, e.g. `address.city`. By default it's also the name of the expected HTTP parameter, e.g. `GET http://myhost?address.city=Springfield`.
 
+For parameter with multiple values property `paramSeparator` could be specified. The value of `paramSeparator` is a regular expression used to param delimitation. For example if param separator is specified :`\\,`, parameter: `?param=val1,val2,val3` will be processed as multi value param: `["val1","val2","val3"]`, by default (without specified `paramSeparator`) param will be processed as single value : `"val1,val2,val3"`.
```
Comment (tkaczmarzyk) [inline README.md:94]: 1) get rid of regexps, please
2) I think we need to describe the standard way as well (&name=Homer&name=Moe ....) and clearly show that both are supported

File: README.md [hunk 4]
```diff
@@ -133,15 +135,21 @@ A negation for this specification is also available: `NotEqualIgnoreCase`.
...
+Supports multiple data types: numbers, booleans, strings, dates, enums.
+
+Usage: `@Spec(path="gender", spec=In.class)`.
+
 HTTP request example:
 
     GET http://myhost/customers?gender=MALE&gender=FEMALE
 
-Supports multiple data types: numbers, booleans, strings, dates, enums.
+or if `paramSeparator` is specified (eg. `@Spec(path="gender", paramSeparator=",", spec=In.class)`):
```
Comment (tkaczmarzyk) [inline README.md:146]: as you can see, it's easy to forget regexp, you have "," here, and "\\," above :)

File: src/main/java/net/kaczmarzyk/spring/data/jpa/web/SimpleSpecificationResolver.java [hunk 5]
```diff
@@ -145,11 +145,19 @@ private Converter resolveConverter(Spec def) {
...
 	private Collection<String> resolveSpecArgumentsFromHttpParameters(WebRequestProcessingContext context, Spec specDef) {
 		Collection<String> args = new ArrayList<String>();
-		
+
 		if (specDef.params().length != 0) {
+			String paramsSeparator = specDef.paramSeparator();
 		    for (String webParamName : specDef.params()) {
 		        String[] httpParamValues = context.getParameterValues(webParamName);
-		        addValuesToArgs(httpParamValues, args);
+			    if(paramsSeparator.isEmpty()) {
```
Comment (tkaczmarzyk) [inline src/main/java/net/kaczmarzyk/spring/data/jpa/web/SimpleSpecificationResolver.java:153]: space after `if`, please

File: src/main/java/net/kaczmarzyk/spring/data/jpa/web/SimpleSpecificationResolver.java [hunk 6]
```diff
@@ -145,11 +145,19 @@ private Converter resolveConverter(Spec def) {
...
+				    addValuesToArgs(httpParamValues, args);
+			    } else {
+			    	for(String paramValue: httpParamValues) {
```
Comment (tkaczmarzyk) [inline src/main/java/net/kaczmarzyk/spring/data/jpa/web/SimpleSpecificationResolver.java:156]: space after `for` and before `:` please

File: src/main/java/net/kaczmarzyk/spring/data/jpa/web/annotation/Spec.java [hunk 7]
```diff
@@ -32,7 +32,12 @@
 public @interface Spec {
 
     String[] params() default {};
-    
+
+    /**
```
Comment (tkaczmarzyk) [inline src/main/java/net/kaczmarzyk/spring/data/jpa/web/annotation/Spec.java:36]: please make the implementation non-regexp

File: src/test/java/net/kaczmarzyk/InE2eTest.java [hunk 8]
```diff
@@ -37,104 +37,184 @@
...
 	public void findsByListOfAllowedStringValues() throws Exception {
 		mockMvc.perform(get("/customers")
 				.param("firstNameIn", "Homer", "Marge")
 				.accept(MediaType.APPLICATION_JSON))
-			.andExpect(status().isOk())
-			.andExpect(jsonPath("$").isArray())
-			.andExpect(jsonPath("$[0].firstName").value("Homer"))
-			.andExpect(jsonPath("$[1].firstName").value("Marge"))
-			.andExpect(jsonPath("$[2]").doesNotExist());
+				.andExpect(status().isOk())
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/InE2eTest.java:113]: this formmating seems werid to me, the `andExpect` is different level of invocation `()` than `accept`

File: src/test/java/net/kaczmarzyk/NotInE2eTest.java [hunk 9]
```diff
@@ -101,55 +148,99 @@ public void findsByListOfNotAllowedLongValues() throws Exception {
...
+
+	@Test
+	public void findsByListOfNotAllowedLongValuesUsingParamSeparator() throws Exception {
+		String idNotInParam = "?idNotIn="
+				+ homerSimpson.getId().toString() + "_"
+				+ moeSzyslak.getId().toString() + "_"
+				+ margeSimpson.getId().toString() + "_"
+				+ nedFlanders.getId().toString();
+
+		mockMvc.perform(get("/not-in/customers-ps"+idNotInParam)
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/NotInE2eTest.java:167]: spaces arount `+`, please

Comment (tkaczmarzyk) [reply src/test/java/net/kaczmarzyk/NotInE2eTest.java:167]: "ps" -> "paramSeparator" for better readability

File: src/test/java/net/kaczmarzyk/NotInE2eTest.java [hunk 10]
```diff
@@ -101,55 +148,99 @@ public void findsByListOfNotAllowedLongValues() throws Exception {
...
+				.andExpect(jsonPath("$").isArray())
+				.andExpect(jsonPath("$[0].firstName").value("Marge"))
+				.andExpect(jsonPath("$[1].firstName").value("Bart"))
+				.andExpect(jsonPath("$[2].firstName").value("Lisa"))
+				.andExpect(jsonPath("$[3].firstName").value("Maggie"))
+				.andExpect(jsonPath("$[4].firstName").value("Moe"))
+				.andExpect(jsonPath("$[5].firstName").value("Ned"))
+				.andExpect(jsonPath("$[6]").doesNotExist());
 	}
 }
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/NotInE2eTest.java:246]: General note: you are missing a test which would check that when paramSeparator is not specified, then values such as "MALE,FEMALE" will not be handled as a multiple values. This should be tested at unit level (SimpleSpecificationResolver) and in one e2e test level (actual controller without paramSeparator specified)

Comment (tkaczmarzyk) [reply src/test/java/net/kaczmarzyk/NotInE2eTest.java:246]: When you remove regex support, I would explicitly test it too

File: src/main/java/net/kaczmarzyk/spring/data/jpa/web/SimpleSpecificationResolver.java [hunk 11]
```diff
@@ -146,39 +153,67 @@ private Converter resolveConverter(Spec def) {
...
+		if (paramValues != null) {
+			for (String paramValue : paramValues) {
+				if (!StringUtils.isEmpty(paramValue)) {
+					args.add(paramValue);
+				}
+			}
+		}
+	}
+
+	private static class DelimitationPattern {
```
Comment (tkaczmarzyk) [inline src/main/java/net/kaczmarzyk/spring/data/jpa/web/SimpleSpecificationResolver.java:192]: Consider renaming:
DelimitationPattern -> DelimitationStrategy
EMPTY -> NONE

File: src/test/java/net/kaczmarzyk/InE2eTest.java [hunk 12]
```diff
@@ -110,111 +126,151 @@ public void findsByListOfAllowedStringValues() throws Exception {
...
+			.andExpect(jsonPath("$[2]").doesNotExist());
+	}
+
+	@Test
+	public void findsByListOfAllowedStringValuesUsingWebParameterNameTheSameAsPathNameWithoutUsingParamSeparator() throws Exception {
+		mockMvc.perform(get("/customers-with-missing-parameter-name?firstName=Homer&firstName=Moe,Bart&firstName=Lisa,Maggie")
+				.accept(MediaType.APPLICATION_JSON))
+			.andExpect(status().isOk())
+			.andExpect(jsonPath("$").isArray())
+			.andExpect(jsonPath("$[0].firstName").value("Homer"))
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/InE2eTest.java:168]: I don't get it why only Homer is here. If delimitation is disabled, I would expect these values: "Homer", "Moe,Bart", "Lisa,Maggie" 

??

File: src/test/java/net/kaczmarzyk/InE2eTest.java [hunk 13]
```diff
@@ -110,111 +126,151 @@ public void findsByListOfAllowedStringValues() throws Exception {
...
+			.andExpect(jsonPath("$[2]").doesNotExist());
+	}
+
+	@Test
+	public void findsByListOfAllowedStringValuesUsingWebParameterNameTheSameAsPathNameAndParamSeparator() throws Exception {
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/InE2eTest.java:173]: findsByListOfAllowedStringValuesUsingWebParameterNameTheSameAsPathNameAndParamSeparator

->findsByListOfAllowedStringValuesUsing**ParamSeparator**AndWebParameterNameTheSameAsPathName

File: src/test/java/net/kaczmarzyk/InE2eTest.java [hunk 14]
```diff
@@ -110,111 +126,151 @@ public void findsByListOfAllowedStringValues() throws Exception {
...
+			.andExpect(jsonPath("$").isArray())
+			.andExpect(jsonPath("$[0].firstName").value("Homer"))
+			.andExpect(jsonPath("$[1].firstName").value("Moe"))
+			.andExpect(jsonPath("$[2]").doesNotExist());
 	}
 
 	@Test
 	public void findsByListOfAllowedLongValuesUsingParamSeparator() throws Exception {
-		mockMvc.perform(get("/customers-ps?idIn=" + homerSimpson.getId().toString() + ";" + moeSzyslak.getId().toString())
+		mockMvc.perform(get("/customers-param-separator?idIn=" + homerSimpson.getId().toString() + ";" + moeSzyslak.getId().toString())
```
Comment (tkaczmarzyk) [inline src/test/java/net/kaczmarzyk/InE2eTest.java:200]: Is it possible to use
```
.param("idIn", homerSimpson.getId().toString() + ";" + moeSzyslak.getId().toString())

File: src/test/java/net/kaczmarzyk/InE2eTest.java (hunk 12 above)
Comment (jradlica) [reply src/test/java/net/kaczmarzyk/InE2eTest.java:168]: because there are not any customer with firstName `Moe,Bart` or `Lisa,Maggie` does not exist

**Pull request merged**